### PR TITLE
Feature/paginated

### PR DESCRIPTION
**Pull request by FabianEllenberger**
enabled paginated posts fetching

### CODE REVIEW

**Comment by FabianEllenberger** [comment]
enabled paginated posts fetching

**Comment by marcoeh** [comment]
@d-simon plz review :v:

**Comment by FabianEllenberger** [comment]
@d-simon updated some namings, what do you think?

File: README.md [hunk 1]
```diff
@@ -37,6 +37,7 @@ const config = {
   savePath: './static/data',
   compressJSON: true, // setting this to false may help debugging :-)
   perPage: 5000, // arbitrary
+  postsPerPage: 10, // set for paginated posts (default will automatically be 10)
```
Comment (marcoeh) [inline README.md]: post = wp post object (post, page, etc.)
or
post = wp post type `post` (if yes, maybe use `itemsPerPage`

Comment (marcoeh) [reply README.md]: anyway check if the example is correct in readme :)

Comment (FabianEllenberger) [reply README.md]: 👍

File: src/lib/paginate.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+       saveDataToFile(data)
+       log.success('DONE.')
+     })
+   }
+
+   /**
+    * Save files (in this case paginated)
+    */
+  function saveDataToFile (data) {
+    const dataClone = cloneDeep(data) // deep clone the data in order to do calculations
```
Comment (d-simon) [inline src/lib/paginate.js]: Why is the cloning needed?  This adds a new dependency `lodash.clonedeep`, but we only ever seem to read the length of it?

File: src/lib/bundle.js [hunk 3]
```diff
@@ -24,31 +24,22 @@ export default function bundle (bundleName, fetchOptions, _config) {
...
-      function (err) {
-        if (err) return console.error(err)
-        log.success(`Wrote to ${config.savePath}/${fileName} successfully!`)
-      }
-    )
+  /**
+   * Save files (in this case as a bundle)
+   */
+  function saveDataToFile (data) {
+    saveFiles(data, bundleName, config)
```
Comment (d-simon) [inline src/lib/bundle.js]: Should return:  `return saveFiles(…)`
Because saveFiles might return a promise (or whatever)

File: src/lib/bundle.js [hunk 4]
```diff
@@ -24,31 +24,22 @@ export default function bundle (bundleName, fetchOptions, _config) {
...
-      datas.forEach(saveFile)
+      // for each lang save files
+      datas.forEach(saveDataToFile)
       log.success('DONE.')
     })
   } else {
     fetchData(config, false, fetchOptions).then((data) => {
-      saveFile(data)
+      // if one language, only save this lang
+      saveDataToFile(data)
```
Comment (d-simon) [inline src/lib/bundle.js]: This runs async, so we should use a callback.
```js
saveDataToFile(data).then(()=>log.success('DONE.')
```

File: src/lib/bundle.js [hunk 5]
```diff
@@ -24,31 +24,22 @@ export default function bundle (bundleName, fetchOptions, _config) {
     Promise.all(
       config.languages.map(language => fetchData(config, language, fetchOptions))
     ).then((datas) => {
-      datas.forEach(saveFile)
+      // for each lang save files
+      datas.forEach(saveDataToFile)
```
Comment (d-simon) [inline src/lib/bundle.js]: This runs asynchronously, which means the log following it is executed before it.
We can use a `map` to fire all asynchronous `saveDataToFile()`-calls and put them in an array. Then we use `Promise.all(arrayOfPromises)` to resolve them.

```js

return Promise.all( // resolves an array of promises
  datas.map((data) => { // returns array of all promises from all saveDataToFile()-calls
    return saveDataToFile(data)
  })
).then(() => log.success('DONE.')) // is called when all promises are resolved (here: all files are saved)
```

File: src/lib/paginate.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+  /**
+   * Fetch all languages
+   */
+   if (config.languages && config.languages.length) {
+     Promise.all(
+       config.languages.map(language => fetchData(config, language, fetchOptions))
+     ).then((datas) => {
+       // for each lang save files
+       datas.forEach(saveDataToFile)
```
Comment (d-simon) [inline src/lib/paginate.js]: Make async with `map` and `Promise.all` (see annotation in bundle.js)

File: src/lib/paginate.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+       log.success('DONE.')
+     })
+   } else {
+     fetchData(config, false, fetchOptions).then((data) => {
+       // if one language, only save this lang
+       saveDataToFile(data)
```
Comment (d-simon) [inline src/lib/paginate.js]: Make async

File: src/lib/paginate.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+       log.success('DONE.')
+     })
+   }
+
+   /**
+    * Save files (in this case paginated)
+    */
+  function saveDataToFile (data) {
```
Comment (d-simon) [inline src/lib/paginate.js:45]: This function should return a promise which is a `Promise.all()` of an array of all the promises of all files being saved. (got'em?)

File: src/lib/paginate.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    const dataClone = cloneDeep(data) // deep clone the data in order to do calculations
+    const itemCount = _config.itemsPerPage ? _config.itemsPerPage : 10 // how many items per page, default 10
+    const itemsTotal = dataClone[bundleName].length // itemsTotal - items length
+    const slice = Math.ceil(itemsTotal / itemCount) // round up slices (101 items will be 11 pages – last page with 1 item)
+    let from = 0
+    for (var i = 1; i <= slice; i++) {
+      saveFiles(
```
Comment (d-simon) [inline src/lib/paginate.js]: We need to "collect" all these calls into an array and resolve it. (see annotation in bundle.js)

**Review by d-simon — CHANGES_REQUESTED** [review]
Great stuff 👍 
I am a bit tired, but I think it is largely okay. Only the async handling is something that needs to improve. (Always return a promise)

File: src/lib/bundle.js (hunk 5 above)
Comment (d-simon) [reply src/lib/bundle.js]: This applies for all other changes as well ✨

File: src/lib/paginate.js (hunk 2 above)
Comment (FabianEllenberger) [reply src/lib/paginate.js]: the length changes because we slice items from that array, that's why I cloned it

File: src/lib/saveFiles.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+ * Save files
+ */
+export default function saveFiles (data, bundleName, config, pageNumber) {
+  const json = JSON.stringify(data, null, config.compressJSON ? null : 2)
+  const jsonSizeKB = (Math.round((Buffer.byteLength(json, 'utf8') / 1024) * 100)) / 100
+  const fileName = `${bundleName}${data.language ? `.${data.language}` : ''}${pageNumber ? `.${pageNumber}` : ''}.json`
+
+  log.info(`Writing ${fileName} (Length: ${json.length}, Size: ${jsonSizeKB}kB)`)
+
+  fs.writeFile(
```
Comment (d-simon) [inline src/lib/saveFiles.js]: This should be returned

**Pull request merged**